pkgs/tools/virtualization/mkosi/default.nix

```diff
@@ -38,35 +39,52 @@ let
     withKernelInstall = true;
   };
 
-  python3pefile = python3.withPackages (ps: with ps; [
-    pefile
-  ]);
+  python3pefile = python3.withPackages (
+    ps: with ps; [
+      pefile
+    ]
+  );
 in
 buildPythonApplication rec {
   pname = "mkosi";
-  version = "22";
+  version = "24.3-unstable-2024-08-28";
   format = "pyproject";
 
-  outputs = [ "out" "man" ];
+  outputs = [
+    "out"
+    "man"
+  ];
 
   src = fetchFromGitHub {
     owner = "systemd";
     repo = "mkosi";
-    rev = "v${version}";
-    hash = "sha256-Zom1GlyhqgpTKfjcBOUEJMlubSn+TQsk97js1/UfDHY=";
+    rev = "8c2f828701a1bdb3dc9b80d6f2ab979f0430a6b8";
+    hash = "sha256-rO/4ki2nAJQN2slmYuHKESGBBDMXC/ikGf6dMDcKFr4=";
   };
 
-  # Fix ctypes finding library
-  # https://github.com/NixOS/nixpkgs/issues/7307
-  postPatch = lib.optionalString stdenv.isLinux ''
-    substituteInPlace mkosi/user.py \
-      --replace-fail 'ctypes.util.find_library("c")' "'${stdenv.cc.libc}/lib/libc.so.6'"
-    substituteInPlace mkosi/__init__.py \
-      --replace-fail '/usr/lib/systemd/ukify' "${systemdForMkosi}/lib/systemd/ukify"
-  '' + lib.optionalString withQemu ''
-    substituteInPlace mkosi/qemu.py \
-      --replace-fail "usr/share/qemu/firmware" "${qemu}/share/qemu/firmware"
-  '';
+  postPatch =
+    ''
+      # Fix ctypes finding library
+      # https://github.com/NixOS/nixpkgs/issues/7307
+      substituteInPlace mkosi/__init__.py \
+        --replace-fail '/usr/lib/systemd/ukify' "${systemdForMkosi}/lib/systemd/ukify" \
+        --replace-fail 'python_binary(context.config, binary=ukify),' ""
+      substituteInPlace mkosi/sandbox/__init__.py \
+        --replace-fail 'ctypes.CDLL(None, use_errno=True)' 'ctypes.CDLL("${stdenv.cc.libc}/lib/libc.so.6", use_errno=True)' \
+        --replace-fail 'ctypes.CDLL("libseccomp.so.2")' 'ctypes.CDLL("${libseccomp.lib}/lib/libseccomp.so.2")'
+
+      # Rather than calling mkosi from mkosi with sys.executable, which doesn't use the Python wrappers for PATH
+      # and PYTHONPATH, we call the wrapped binary as an entrypoint.
+      substituteInPlace mkosi/run.py \
+        --replace-fail 'sys.executable, "-SI", mkosi.sandbox.__file__' "\"$out/bin/mkosi-sandbox\"" \
+        --replace-fail '"python3" if tools or not exe.is_relative_to("/usr") else exe, "-SI", "/sandbox.py"' "\"$out/bin/mkosi-sandbox\""
+      substituteInPlace mkosi/__init__.py \
+        --replace-fail '"python3" if (tools and config.tools_tree) or not exe.is_relative_to("/usr") else exe' "\"${python3pefile}/bin/python3.12\""
```

Comment:
I do find these to be clearer and more future-proof -- and since they include the context, more reviewable -- as a patch file.

Reply:
If a patch - we'd have to pre-process it with a placeholder to later replace with the actual dependency. Or is there another way? I'm actually all on your side here, but didn't have an apparent solution to this in mind

Reply:
Yeah, I'll refactor it to use that tomorrow. Probably still more expressive than what we currently have.

Reply:
Refactored that :)